# **1. PRE-PROCESSING DATA**

### **Các bước tiền xử lý áp dụng:**
1. Kiểm tra giá trị trùng lặp
2. Xử lý giá trị thiếu
3. Làm sạch và Chuẩn hóa dữ liệu
4. Đồng bộ hóa tên cột

## Import thư viện cần thiết

In [470]:
import pandas as pd
import numpy as np
import re
import seaborn as sns
import matplotlib.pyplot as plt
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.metrics.pairwise import cosine_similarity

## Đọc dữ liệu

In [471]:
df = pd.read_csv("../data/youtube_vn_music.csv")
print("Tổng quan dữ liệu:")
df.info()

Tổng quan dữ liệu:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10615 entries, 0 to 10614
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   video_id                  10615 non-null  object 
 1   published_at              10615 non-null  object 
 2   channel_id                10615 non-null  object 
 3   channel_title             10615 non-null  object 
 4   title                     10615 non-null  object 
 5   description               9536 non-null   object 
 6   tags                      8725 non-null   object 
 7   category_id               10615 non-null  int64  
 8   view_count                10615 non-null  int64  
 9   like_count                10018 non-null  float64
 10  comment_count             10451 non-null  float64
 11  duration                  10615 non-null  object 
 12  definition                10615 non-null  object 
 13  caption                   10615 non-null  

In [472]:
df.head(5)

,video_id,published_at,channel_id,channel_title,title,description,tags,category_id,view_count,like_count,...,caption,licensed_content,privacy_status,made_for_kids,tags_count,topic_categories,search_query,channel_subscriber_count,channel_view_count,channel_video_count
0,sZrIbpwjTwk,2025-08-17T05:00:48Z,UC2fu6CiFfNYz5UFORvFyc0w,Vie Channel - MUSIC,ẾCH NGOÀI ĐÁY GIẾNG - Phương Mỹ Chi hoá thân thành em ếch xinh | Em Xinh Say Hi [Performance],ẾCH NGOÀI ĐÁY GIẾNG - Phương Mỹ Chi hoá thân thành em ếch xinh | Em Xinh Say Hi [Performance] \n\n---\n#emxinhsayhi #emxinh #echngoaidaygieng #phuongmychi #EmXinhSayHi #DauTruongGiaToc\n#viechannelmusic #VieChannelMUSIC #VieChannel #VieON #DatVietVAC \n#30NamCoMotDatViet \n\n👉 Xem ngay tập full: https://www.youtube.com/watch?v=O-_lQxbES_s\n👉 Xem sớm nhất trên ứng dụng VieON: https://vieon.vn/phim-hay/m/tv-show-game-thuc-te/\n👉 Xem thêm tại đây: https://exsh.bfan.link/EXSH-EP12\n\nĐăng ký hội viên ngay nào👇:\nhttps://www.youtube.com/channel/UCkna2OcuN1E6u5I8GVtdkOw/join\n\n★ Mọi nhu cầu quảng cáo và tài trợ xin liên hệ: Công Ty Cổ Phần Vie Channel\n► Hotline: 0907.950.371 (Mr. Đức Khoa)\n► Email: khoa.nguyen@viechannel.vn hoặc sales@viechannel.vn\n\n📣 Đừng quên subscribe để thưởng thức thật nhiều bài hát hay nhất nhé! \n🎈 Subscribe ngay: https://www.youtube.com/channel/UC2fu6CiFfNYz5UFORvFyc0w?sub_confirmation=1\n\nTải app VieON tại: https://click.vieon.vn/wSBG/ytapp\n\nXem các chương trình hấp dẫn khác của Vie Channel:\n►Em Xinh Say Hi Full: https://bit.ly/EmXinhSayHi2025\n►Em Xinh Say Hi Performance: https://bit.ly/EmXinhSayHiPerformance\n►Anh Trai Say Hi Full: https://bit.ly/AnhTraiSayHi\n►Anh Trai Say Hi Performance: https://bit.ly/AnhTraiSayHiPerformance\n►Rap Việt Full: https://bit.ly/RAPVIETSS3\n►Rap Việt Performance: https://bit.ly/RapViet2024Performance\n\nHoặc xem thêm trên hệ thống Vie Channel:\n►Website: http://viechannel.vn/\n►Fanpage: https://www.facebook.com/VieChannelHTV2/\n►Instagram: https://www.instagram.com/viechannelhtv2/,"htv2|vie channel|music vie channel|em xinh say hi|Em Xinh Say Hi|vie channel em xinh say hi|Em Xinh ""Say Hi"" 2025|em xinh say hi 2025 mới nhất|em xinh say hi tập 12|tập 12 em xinh|tập 12 em xinh say hi|em xinh tập 12|livestage 5 em xinh say hi|solo|solo em xinh say hi|ếch ngoài đáy giếng|ếch ngoài đáy giếng phương mỹ chi|phương mỹ chi|phương mỹ chi ếch ngoài đáy giếng|ếch ngoài đáy giếng live performance|ếch ngoài đáy giếng solo performance|phuong my chi",10,82407446,234808.0,...,False,True,public,False,22,https://en.wikipedia.org/wiki/Music|https://en.wikipedia.org/wiki/Music_of_Asia|https://en.wikipedia.org/wiki/Pop_music|https://en.wikipedia.org/wiki/Television_program,bài,1710000.0,2.303825e+09,2519.0
1,Qgzb0musH3A,2022-08-24T12:00:12Z,UCIKpxDD9CUS7vBCJte7l-UQ,Như Hoa Official,Bài hát đang Hót🔥TRỞ VỀ_Như Hoa|Bài Hát Lấy Đi Nước Mắt Hàng Triệu Người Con Xa Quê|Vu Lan Báo Hiếu,Bài Hát Đang Hót 🔥 TRỞ VỀ | Như Hoa ( Thần Tượng Bolero 2019) Bài Hát Lấy Đi Nước Mắt Hàng Triệu Người Con Xa Quê | Vu Lan Báo Hiếu 2022. \n[ MV Official]\nComposer: Tuấn Phương.\nSinger: Như Hoa ( Thần tượng Bolero 2019)\nRecord - Mix : Võ Hoàng Studio\nDirectors: Như Hoa Official\nProduced by: Khang Hy Entertainment\n\n ----------------------------\nLyric : TRỞ VỀ.\nSáng tác : Nhạc Sĩ Tuấn Phương.\n❤️\nTa trở về trong tiếng gọi mẹ ơi!\nVề với mẹ về nơi thật yên bình\nNước mắt ta rơi bởi một điều có thật\nRồi một ngày mẹ sẽ đi xa.\n\nTa trở về bên cánh đồng quê xưa\nHạt lúa vàng hạt mồ hôi chát mặn\nTa lớn khôn thêm giữa nghĩa tình xóm làng\nChia cho nhau ngụm nước giữa trưa hè.\nĐK:\nTa về đây khu vườn rộn tiếng ve\nCó cánh diều mang tuổi thơ khát vọng\nCó cánh cò bên nôi mẹ hát\nCó hàng tre ru chiều gió thênh thang.\n\nNếu cuộc sống ban cho ta điều ước\nTa xin được về những năm tháng tuổi thơ\nNghe lá hát trên từng con đường nhỏ\nTổ quốc mình chẳng thể nào thay.\n\nVẫn còn đây tím chiều nhuộm áo em\nVẫn câu hò như lời ai nhắn nhủ\nVẫn con đò mang bao niềm nhớ\nVẫn mẹ ta bên t

## Bước 1: Kiểm tra giá trị trùng lặp

Để kiểm soát *chất lượng* và *số lượng* dữ liệu một cách chặt chẽ hơn, việc đầu tiên cần thực hiện là kiểm tra các bản ghi trùng lặp trong dataset.

Do API có thể lấy về các bản ghi bị lặp lại (ví dụ: do thực hiện nhiều request trong cùng một khoảng thời gian, hoặc API trả về dữ liệu có timestamp trùng nhau), nên việc kiểm tra trùng lặp là bước quan trọng để đảm bảo chất lượng dữ liệu đầu vào.

Việc xử lý trùng lặp ngay từ đầu giúp tránh sai lệch khi phân tích, đồng thời giúp hệ thống quản lý dữ liệu hiệu quả hơn trong các bước tiếp theo.

In [473]:
print(f"Tổng số video: {len(df)}")
print(f"Video có ids độc nhất: {df['video_id'].nunique()}")
print(f"Video có ids bị trùng: {len(df) - df['video_id'].nunique()}")

Tổng số video: 10615
Video có ids độc nhất: 7376
Video có ids bị trùng: 3239


Dữ liệu có xảy ra trùng lặp video ID, với số lượng video bị trùng là 192 video.

==> Cần loại bỏ các bản ghi trùng lặp này để đảm bảo tính duy nhất của mỗi video trong dataset.

In [474]:
# Loại bỏ các bản ghi trùng lặp dựa trên video_id
df = df.drop_duplicates(subset=['video_id'])
df = df.reset_index(drop=True)
df.info()
print("Sau khi loại bỏ trùng lặp:")
print(f"Tổng số video: {len(df)}")
print(f"Video có ids độc nhất: {df['video_id'].nunique()}")
print(f"Video có ids bị trùng: {len(df) - df['video_id'].nunique()}")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7376 entries, 0 to 7375
Data columns (total 23 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   video_id                  7376 non-null   object 
 1   published_at              7376 non-null   object 
 2   channel_id                7376 non-null   object 
 3   channel_title             7376 non-null   object 
 4   title                     7376 non-null   object 
 5   description               6689 non-null   object 
 6   tags                      6071 non-null   object 
 7   category_id               7376 non-null   int64  
 8   view_count                7376 non-null   int64  
 9   like_count                6880 non-null   float64
 10  comment_count             7264 non-null   float64
 11  duration                  7376 non-null   object 
 12  definition                7376 non-null   object 
 13  caption                   7376 non-null   bool   
 14  licensed

## Bước 2: Xử lý giá trị thiếu

Dựa trên kết quả `df.info()` ở bước đọc dữ liệu, ta đã quan sát sơ bộ rằng dataset có xuất hiện giá trị thiếu ở một số cột.

Để có cái nhìn cụ thể và chính xác hơn, ta sẽ tiến hành:
- Thống kê số lượng giá trị thiếu trên từng cột.
- Xác định tỷ lệ phần trăm giá trị thiếu so với tổng số bản ghi.
- Đề xuất hướng xử lý phù hợp cho từng loại cột.

In [475]:
# Kiểm tra giá trị thiếu
missing_values = df.isnull().sum()
print("Giá trị thiếu trong mỗi cột:")
print(missing_values[missing_values > 0])

# Tính tỷ lệ phần trăm giá trị thiếu
missing_percentage = (missing_values / len(df)) * 100
print("Tỷ lệ phần trăm giá trị thiếu trong mỗi cột:")
print(missing_percentage[missing_percentage > 0])

Giá trị thiếu trong mỗi cột:
description                  687
tags                        1305
like_count                   496
comment_count                112
topic_categories              15
channel_subscriber_count       2
channel_view_count             2
channel_video_count            2
dtype: int64
Tỷ lệ phần trăm giá trị thiếu trong mỗi cột:
description                  9.313991
tags                        17.692516
like_count                   6.724512
comment_count                1.518438
topic_categories             0.203362
channel_subscriber_count     0.027115
channel_view_count           0.027115
channel_video_count          0.027115
dtype: float64


Có 8 cột có giá trị thiếu:
- `description`
- `tags`
- `like_count`
- `comment_count`
- `topic_categories`
- `channel_subscriber_count`
- `channel_view_count`
- `channel_video_count`

### **Xử lý `description`**

`description` là phần mô tả nội dung video do người tạo video tự cung cấp. Đây không phải là trường dữ liệu bắt buộc và việc thiếu mô tả không ảnh hưởng đến các phân tích định lượng chính của dataset.

Do đó, với các bản ghi bị thiếu giá trị ở cột `description`, ta lựa chọn điền giá trị rỗng ("") để giữ nguyên số lượng bản ghi và đảm bảo các xử lý văn bản (nếu thực hiện sau này) có thể nhận diện dễ dàng trường hợp mô tả trống.

In [476]:
# Description
df['description'] = df['description'].fillna('')

### **Xử lý `like_count`**

Đối với cột `like_count`, kết quả thống kê cho thấy có 496 dòng bị thiếu (chiếm ~6.72% trên toàn tập dữ liệu). 

Nhận thấy thuộc tính này là một trong những thuộc tính quan trọng của quá trình phân tích và trong đồ án này nhóm sẽ tập trung vào những dữ liệu chất lượng, đầy đủ insight để phân tích ở phần sau. Do đó, thay vì gán giá trị mặc định hoặc nội suy (vốn không phù hợp với dữ liệu dạng thời gian có tính chất duy nhất cho từng kênh), ta quyết định **loại bỏ 496 dòng dữ liệu** thiếu ở trường này.

In [477]:
df = df.dropna(subset=['like_count'])

Sau khi loại bỏ 496 dòng `like_count`:

In [478]:
# Kiểm tra giá trị thiếu
missing_values = df.isnull().sum()
print("Giá trị thiếu trong mỗi cột:")
print(missing_values[missing_values > 0])

# Tính tỷ lệ phần trăm giá trị thiếu
missing_percentage = (missing_values / len(df)) * 100
print("Tỷ lệ phần trăm giá trị thiếu trong mỗi cột:")
print(missing_percentage[missing_percentage > 0])

Giá trị thiếu trong mỗi cột:
tags                        1232
comment_count                106
topic_categories              15
channel_subscriber_count       1
channel_view_count             1
channel_video_count            1
dtype: int64
Tỷ lệ phần trăm giá trị thiếu trong mỗi cột:
tags                        17.906977
comment_count                1.540698
topic_categories             0.218023
channel_subscriber_count     0.014535
channel_view_count           0.014535
channel_video_count          0.014535
dtype: float64


### **Xử lý `comment_count`**

Đối với cột `comment_count`, kết quả thống kê cho thấy có 106 dòng bị thiếu sau khi đã xử lý `like_count` (chiếm ~1.54% trên toàn tập dữ liệu). 

Nhận thấy thuộc tính này là một trong những thuộc tính quan trọng của quá trình phân tích và trong đồ án này nhóm sẽ tập trung vào những dữ liệu chất lượng, đầy đủ insight để phân tích ở phần sau. Do đó, thay vì gán giá trị mặc định hoặc nội suy (vốn không phù hợp với dữ liệu dạng thời gian có tính chất duy nhất cho từng kênh), ta quyết định **loại bỏ 106 dòng dữ liệu** thiếu ở trường này.

In [479]:
df = df.dropna(subset=['comment_count'])

Sau khi loại bỏ 106 dòng `comment_count`:

In [480]:
# Kiểm tra giá trị thiếu
missing_values = df.isnull().sum()
print("Giá trị thiếu trong mỗi cột:")
print(missing_values[missing_values > 0])

# Tính tỷ lệ phần trăm giá trị thiếu
missing_percentage = (missing_values / len(df)) * 100
print("Tỷ lệ phần trăm giá trị thiếu trong mỗi cột:")
print(missing_percentage[missing_percentage > 0])

Giá trị thiếu trong mỗi cột:
tags                1207
topic_categories      14
dtype: int64
Tỷ lệ phần trăm giá trị thiếu trong mỗi cột:
tags                17.818128
topic_categories     0.206673
dtype: float64


### **Xử lý `tags`**

`tags` là trường chứa các thẻ gắn với video, thường được sử dụng để mô tả chủ đề hoặc giúp thuật toán phân loại nội dung tốt hơn. Đây là một trong những loại dữ liệu có thể mang ý nghĩa khi phân tích chủ đề, phân cụm nội dung hoặc xây dựng mô hình gợi ý. Tuy nhiên, thống kê cho thấy cột này có tỷ lệ giá trị thiếu khá lớn.

Do mức độ thiếu nhiều và chưa có đủ thông tin để xây dựng phương án xử lý phù hợp (ví dụ: suy đoán dựa trên tiêu đề, mô tả hoặc nội dung video), ta tạm thời áp dụng cách tiếp cận an toàn là điền giá trị rỗng ("") cho các trường thiếu.

In [481]:
df['tags'] = df['tags'].fillna('')

### **Xử lý `topic_categories`**

Cột `topic_categories` thể hiện các chủ đề mà video thuộc về, thường được gắn tự động bởi hệ thống phân loại nội dung. Trường thông tin này có thể hữu ích trong các phân tích về chủ đề, xu hướng hoặc nhóm nội dung. 

Do mức độ thiếu là tương đối ít khi so với toàn bộ dữ liệu (14 dòng - ~0.2%), ta lựa chọn phương án đơn giản và an toàn: điền giá trị rỗng ("") cho các bản ghi thiếu `topic_categories`.

In [482]:
df['topic_categories'] = df['topic_categories'].fillna('')

## Bước 3: Làm sạch và Chuẩn hóa dữ liệu

Sau khi xử lý các vấn đề về trùng lặp và giá trị thiếu, bước tiếp theo là làm sạch và chuẩn hóa dữ liệu nhằm đảm bảo tính nhất quán, đồng nhất và dễ xử lý trong các bước phân tích sau.

Trong bước này, ta thực hiện:
- Chuẩn hóa kiểu dữ liệu (datetime, số nguyên, số thực, chuỗi văn bản, v.v.).
- Làm sạch các cột văn bản.
- Loại bỏ các thuộc tính lỗi/nhiễu/không có giá trị sử dụng.

Do tập dữ liệu bao gồm 19 thuộc tính, ta sẽ chia theo trường dữ liệu để **Làm sạch và Chuẩn hóa**:
- IDs: Các trường có chức năng định danh (`video_id`, `channel_id`, `category_id`).
- Numeric: (`view_count`, `like_count`, `comment_count`, ...).
- Text: (`title`, `description`, `tags`, `topics`, ...).
- Binary: True/False (`caption_status`)
- Datetime: (`published_at`, `channel_created_at`, ...).

### **Xử lý IDs**

Ta sẽ tiến hành kiểm tra **IDs** theo quy trình:
- Xem dữ liệu.
- Làm sạch - Chuẩn hóa dữ liệu.

In [483]:
id_cols = ['video_id', 'channel_id', 'category_id']
print("Các cột ID:", id_cols)

Các cột ID: ['video_id', 'channel_id', 'category_id']


### `video_id`, `channel_id`

In [484]:
print("10 giá trị khác nhau trong cột video_id:")
print(df['video_id'].unique()[:10])

print("\n")

print("10 giá trị khác nhau trong cột channel_id:")
print(df['channel_id'].unique()[:10])

10 giá trị khác nhau trong cột video_id:
['sZrIbpwjTwk' 'Qgzb0musH3A' 'FxX5pTGW2dQ' 'QzwCfC5x-Qs' 'Ku9gOe87Ah0'
 'znkyGyuroMQ' 'u_PCv-ugyR8' '6PV-2d40Fr8' 'aW3qoZeqHik' 'aXQvG0z8up8']


10 giá trị khác nhau trong cột channel_id:
['UC2fu6CiFfNYz5UFORvFyc0w' 'UCIKpxDD9CUS7vBCJte7l-UQ'
 'UCbwr_JmPZbfYsHvtWR0j6bw' 'UCmzOQExdzXzBAbnCbK3qaGA'
 'UCXqXffFOpEkNC8Yl7YEr6Vw' 'UCHKcuuAYs3c8n990WeMs-Lw'
 'UCADqthTjLtL6siuuZ_G56eA' 'UC6eq3sR4CtbvGdmInchzWUA'
 'UCtT5uP2WIGxyP0MYCC_ZHtQ' 'UCq_28uJmbq4tlji7YAMZGtg']



==>> Các cột `video_id` và `channel_id` dùng để định danh duy nhất cho từng video và kênh, giúp phân biệt các bản ghi trong dataset. Vì đây là dữ liệu định danh, không phục vụ trực tiếp cho phân tích định lượng hay mô hình hóa.

Tuy nhiên, ta sẽ giữ lại để phục vụ cho các bước xử lý sau này

### `category_id`

Khác với `video_id` và `channel_id` chỉ dùng để định danh, cột category_id mang ý nghĩa phân loại nội dung của video theo từng chủ đề (ví dụ: Âm nhạc, Giáo dục, Giải trí…) dựa trên [`videoCategories.list` của **YouTube Data API v3**](https://gist.github.com/dgp/1b24bf2961521bd75d6c) (công cụ nhóm sử dụng để thu thập data từ Youtube).

Đây là một đặc trưng quan trọng.

In [485]:
print("10 giá trị khác nhau trong cột category_id:")
print(df['category_id'].unique()[:10])

10 giá trị khác nhau trong cột category_id:
[10]


==>> Do đây `category_id` được lưu dưới dạng `int64` nên ta sẽ xử lý chung với các **cột Numeric cols**.

### **Xử lý Numeric cols (Cột số)**

Ở bước này, các cột dạng số như `view_count`, `like_count`, `comment_count`,... sẽ được làm sạch và chuẩn hóa. Mục tiêu là đảm bảo các giá trị số:
- Đúng kiểu dữ liệu (int64 hoặc float64).
- Loại bỏ giá trị bất thường (kiểm tra và xử lý các outlier hoặc giá trị âm không hợp lý).

In [486]:
numeric_cols = ['category_id', 'view_count', 'like_count', 'comment_count', 'tags_count', 'channel_subscriber_count', 'channel_view_count', 'channel_video_count']
print("Các cột numeric:", numeric_cols)

Các cột numeric: ['category_id', 'view_count', 'like_count', 'comment_count', 'tags_count', 'channel_subscriber_count', 'channel_view_count', 'channel_video_count']


**1. Kiểm tra dữ liệu không phải số (NaN)**

In [487]:
#kiểm tra NaN
nan_counts = df[numeric_cols].isna().sum()
print("Số lượng giá trị NaN trong các cột numeric:")
print(nan_counts)

Số lượng giá trị NaN trong các cột numeric:
category_id                 0
view_count                  0
like_count                  0
comment_count               0
tags_count                  0
channel_subscriber_count    0
channel_view_count          0
channel_video_count         0
dtype: int64


==>> Không có giá trị NaN cần xử lý.

In [488]:
# Kiểm tra giá trị Numeric
print("Mô tả thống kê các cột numeric:")
df[numeric_cols].describe()

Mô tả thống kê các cột numeric:


,category_id,view_count,like_count,comment_count,tags_count,channel_subscriber_count,channel_view_count,channel_video_count
count,6774.0,6.774000e+03,6.774000e+03,6.774000e+03,6774.000000,6.774000e+03,6.774000e+03,6774.000000
mean,10.0,4.174388e+06,2.081897e+04,1.416103e+03,14.742693,7.188978e+05,4.534866e+08,900.408621
std,0.0,1.353087e+07,7.900283e+04,3.118181e+04,9.975864,1.955564e+06,1.651342e+09,1654.260038
min,10.0,0.000000e+00,0.000000e+00,0.000000e+00,0.000000,0.000000e+00,0.000000e+00,1.000000
25%,10.0,1.395325e+04,1.560000e+02,3.000000e+00,6.000000,2.720000e+04,8.781943e+06,158.000000
50%,10.0,2.920230e+05,1.656500e+03,4.500000e+01,16.000000,1.700000e+05,6.154305e+07,427.000000
75%,10.0,2.328462e+06,1.023650e+04,3.017500e+02,22.000000,4.890000e+05,2.172706e+08,945.000000
max,10.0,3.074998e+08,2.655980e+06,2.276298e+06,47.000000,2.000000e+07,1.450046e+10,25999.000000


### **Xử lý Category/Text cols (Cột chữ)**

Các cột chữ bao gồm các cột văn bản (text) như `title`, `description`, `tags`, `topics`. Nhóm này cần được làm sạch và chuẩn hóa để đảm bảo dữ liệu nhất quán và dễ khai thác.

Các bước xử lý chính:
- Làm sạch cột văn bản (Text).
- Kiểm tra độ trùng lặp và thống nhất.
- Vector hóa văn bản & Tính Cosine Similarity.

**Xử lý `definition`**

In [489]:
# kiểm tra unique definition
print(df['definition'].unique())

['hd' 'sd']


- 'hd' - High Definition - Độ nét cao (thường là 720p, 1080p trở lên)
- 'sd' - Standard Definition - Độ nét tiêu chuẩn (thường là 360p, 480p)

Hiện tại cột này không cần xử lý thêm

**Xử lý `privacy_status`**

In [490]:
# kiểm tra cột privacy_status
print(df['privacy_status'].unique())

['public']


Giá trị thu được là dễ hiểu cho ta chỉ có thể thu thập những video ở chế độ **public** khi sử dụng API và query. Vì vậy, trên dữ liệu này thì cột không có giá trị phân tích. Ta sẽ loại bỏ cột này

In [491]:
#bỏ cột privacy_status
df.drop(columns=['privacy_status'], inplace=True)

Kiểm tra lại dữ liệu sau khi drop cột `privacy_status`:

In [492]:
df.info()

<class 'pandas.core.frame.DataFrame'>
Index: 6774 entries, 0 to 7375
Data columns (total 22 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   video_id                  6774 non-null   object 
 1   published_at              6774 non-null   object 
 2   channel_id                6774 non-null   object 
 3   channel_title             6774 non-null   object 
 4   title                     6774 non-null   object 
 5   description               6774 non-null   object 
 6   tags                      6774 non-null   object 
 7   category_id               6774 non-null   int64  
 8   view_count                6774 non-null   int64  
 9   like_count                6774 non-null   float64
 10  comment_count             6774 non-null   float64
 11  duration                  6774 non-null   object 
 12  definition                6774 non-null   object 
 13  caption                   6774 non-null   bool   
 14  licensed_cont

**Xử lý `title`, `description`**

In [493]:
print("5 giá trị khác nhau trong cột title trước khi làm sạch:")
print(df['title'].unique()[:5])

print("\n")

print("5 giá trị khác nhau trong cột description trước khi làm sạch:")
print(df['description'].unique()[:5])

5 giá trị khác nhau trong cột title trước khi làm sạch:
['ẾCH NGOÀI ĐÁY GIẾNG - Phương Mỹ Chi hoá thân thành em ếch xinh | Em Xinh Say Hi [Performance]'
 'Bài hát đang Hót🔥TRỞ VỀ_Như Hoa|Bài Hát Lấy Đi Nước Mắt Hàng Triệu Người Con Xa Quê|Vu Lan Báo Hiếu'
 'Bài Hát ABC - Học Bảng Chữ Cái ABC - Nhạc Thiếu Nhi Remix Sôi Động'
 'Nhạc cuốn là Bé nhảy ngay| Nhảy trend tiktok | #thaytaichannel #shorts #dance'
 'Karaoke Bài Ca Đất Phương Nam Tone Nam Nhạc Sống gia huy karaoke']


5 giá trị khác nhau trong cột description trước khi làm sạch:
['ẾCH NGOÀI ĐÁY GIẾNG - Phương Mỹ Chi hoá thân thành em ếch xinh | Em Xinh Say Hi [Performance] \n\n---\n#emxinhsayhi #emxinh #echngoaidaygieng #phuongmychi #EmXinhSayHi #DauTruongGiaToc\n#viechannelmusic #VieChannelMUSIC #VieChannel #VieON #DatVietVAC \n#30NamCoMotDatViet \n\n👉 Xem ngay tập full: https://www.youtube.com/watch?v=O-_lQxbES_s\n👉 Xem sớm nhất trên ứng dụng VieON: https://vieon.vn/phim-hay/m/tv-show-game-thuc-te/\n👉 Xem thêm tại đây: https://e

In [494]:
def preprocess_text(text):
    if pd.isna(text):
        return []
    
    text = text.lower()
    text = re.sub(r'[^\w\s]', '', text)  # giữ unicode word characters
    return text.split()

df['title'] = df['title'].astype(str)
df['description'] = df['description'].astype(str)
df['title'] = df['title'].apply(lambda x: ' '.join(preprocess_text(x)))
df['description'] = df['description'].apply(lambda x: ' '.join(preprocess_text(x)))

In [495]:
print("5 giá trị khác nhau trong cột title sau khi làm sạch:")
print(df['title'].unique()[:5])

print("\n")

print("5 giá trị khác nhau trong cột description sau khi làm sạch:")
print(df['description'].unique()[:5])

5 giá trị khác nhau trong cột title sau khi làm sạch:
['ếch ngoài đáy giếng phương mỹ chi hoá thân thành em ếch xinh em xinh say hi performance'
 'bài hát đang hóttrở về_như hoabài hát lấy đi nước mắt hàng triệu người con xa quêvu lan báo hiếu'
 'bài hát abc học bảng chữ cái abc nhạc thiếu nhi remix sôi động'
 'nhạc cuốn là bé nhảy ngay nhảy trend tiktok thaytaichannel shorts dance'
 'karaoke bài ca đất phương nam tone nam nhạc sống gia huy karaoke']


5 giá trị khác nhau trong cột description sau khi làm sạch:
['ếch ngoài đáy giếng phương mỹ chi hoá thân thành em ếch xinh em xinh say hi performance emxinhsayhi emxinh echngoaidaygieng phuongmychi emxinhsayhi dautruonggiatoc viechannelmusic viechannelmusic viechannel vieon datvietvac 30namcomotdatviet xem ngay tập full httpswwwyoutubecomwatchvo_lqxbes_s xem sớm nhất trên ứng dụng vieon httpsvieonvnphimhaymtvshowgamethucte xem thêm tại đây httpsexshbfanlinkexshep12 đăng ký hội viên ngay nào httpswwwyoutubecomchanneluckna2ocun1e6u5i8gvtdk

**Xử lý `topic_categories`**

In [496]:
def process_topics_safe(x):
    if pd.isna(x):
        return []
    if isinstance(x, list):
        urls = x
    else:
        # giả sử x là string, có thể chứa nhiều URL phân tách bởi '|'
        urls = x.split('|')
    result = []
    for url in urls:
        if isinstance(url, str):
            # lấy phần cuối URL làm tên topic, loại bỏ '_' thành space
            result.append(url.split('/')[-1].replace('_', ' '))
    return result

df['topic_categories'] = df['topic_categories'].apply(process_topics_safe)

print(df['topic_categories'].head(10))


0               [Music, Music of Asia, Pop music, Television program]
1                                              [Music, Music of Asia]
3                                                             [Music]
5                                      [Hobby, Lifestyle (sociology)]
6                                              [Music, Music of Asia]
7     [Classical music, Music, Music of Asia, Music of Latin America]
9                                                             [Music]
10                                                            [Music]
11                                             [Music, Music of Asia]
12                                  [Action game, Video game culture]
Name: topic_categories, dtype: object


Nhận thấy có 1 topic nằm ngoài phạm trù âm nhạc: Action game, Video game culture. Tiến hành kiểm tra các giá trị thuộc topic này

In [497]:
action_video_game = df[df['topic_categories'].apply(lambda topics: any(t in ['Action game', 'Video game culture'] for t in topics))]
# số lượng giá trị có topic_categories là action game, video game culture
print("Số lượng video có topic_categories là 'Action game' hoặc 'Video game culture': {0}".format(action_video_game.shape[0]))

Số lượng video có topic_categories là 'Action game' hoặc 'Video game culture': 7


In [498]:
action_video_game[['video_id', 'title']].head(7)

,video_id,title
12,aXQvG0z8up8,bài ca sinh nhật tô gà hiha minecraft animation mv
517,QBvL3Ffn1u4,đại chiến rap heroteamofficial x streamers free fire mèo simmy x cô ngân offical mv
549,kU3xP39V6_4,trang phục plants vs zombie phiên bản cặp đôi
2463,0eOPz56PA30,toga tv và nabee đám cưới tvfansinhto fan
2469,A5Q_VTgMXew,không sao cả piano tutorial
2537,jtzhizaCHis,nửa đời tuyết thiên tú x hm remix nhạc phimmv full trong mô tả nhé
4934,GwIiH7czbpU,aut love you lee andy hiha minecraft animation mv


Sau khi tiến hành tìm kiếm và kiểm tra các video này, em nhận thấy có 2/7 videos không liên quan đến âm nhạc. Đó là:
- trang phục plants vs zombie phiên bản cặp đôi
- toga tv và nabee đám cưới tvfansinhto fan

Tiến hành loại bỏ 2 videos này:

In [ ]:
df = df[~df['video_id'].isin(['kU3xP39V6_4', '0eOPz56PA30'])]
df = df.reset_index(drop=True)

**Xử lý `tags`**

Các bước quy trình:
- Chuẩn hóa tags gốc.
- Phân tách dữ liệu.
- Vector hóa văn bản.
- Vector hóa văn bản & Tính Cosine Similarity.
- Gán tags.
- Kiểm tra và lưu giá trị đã xử lý.

In [500]:
print("5 giá trị khác nhau trong cột tags trước khi làm sạch:")
print(df['tags'].unique()[:5])

5 giá trị khác nhau trong cột tags trước khi làm sạch:
['htv2|vie channel|music vie channel|em xinh say hi|Em Xinh Say Hi|vie channel em xinh say hi|Em Xinh "Say Hi" 2025|em xinh say hi 2025 mới nhất|em xinh say hi tập 12|tập 12 em xinh|tập 12 em xinh say hi|em xinh tập 12|livestage 5 em xinh say hi|solo|solo em xinh say hi|ếch ngoài đáy giếng|ếch ngoài đáy giếng phương mỹ chi|phương mỹ chi|phương mỹ chi ếch ngoài đáy giếng|ếch ngoài đáy giếng live performance|ếch ngoài đáy giếng solo performance|phuong my chi'
 'Bài Hát hay nhất về Mẹ|Bài hát ý nghĩa nhất về Mẹ|Bài hát cảm động nhất về mẹ|Vu Lan Báo Hiếu|Như Hoa|Trở Về Như Hoa|Trở Về _ Như Hoa ( Thần Tượng Bolero 2019)|Bài hát đang hót|Trở Về|Như Hoa Bolero|Như Hoa Oficial|Bài hát lấy đi nước mắt hàng triệu người con xa quê|Như Hoa Thần Tượng Bolero 2019|Bài hát đang hót nhất 2022|Mẹ Và Quê Hương'
 'nhạc thiếu nhi|nhạc thiếi|u nhi vui nhộn|nhạc thiếu nhi remix|bai hat thieu nhca nhac thieu nhi|ca nhạc thiếu nhi vui nhộn|nhạc thiếu nhi

**Bước 1:
Chuẩn hóa tags gốc:** Chuyển đổi dữ liệu tags thô từ chuỗi (string) sang danh sách (list), xử lý các dấu phân cách (|, ,) và loại bỏ giá trị rỗng.

In [501]:
def process_tags(x):
    if pd.isna(x):
        return []
    # Nếu đã là list, giữ nguyên
    if isinstance(x, list):
        return [str(t).strip() for t in x if str(t).strip() != '']
    # Nếu là string
    if isinstance(x, str):
        if x == '':
            return []
        if '|' in x:
            return [t.strip() for t in x.split('|') if t.strip() != '']
        elif ',' in x:
            return [t.strip() for t in x.split(',') if t.strip() != '']
        else:
            return [t.strip() for t in x.split() if t.strip() != '']
    return []

df['tags'] = df['tags'].apply(process_tags)

**Bước 2: Phân tách dữ liệu:** Chia tập dữ liệu thành 2 nhóm: nhóm đã có tags (Available) và nhóm bị thiếu tags (Missing).

In [502]:
df_reset = df.reset_index(drop=True)
available_idx = df_reset[df_reset['tags'].map(len) > 0].index
missing_idx = df_reset[df_reset['tags'].map(len) == 0].index

**Bước 3: Vector hóa văn bản & Tính Cosine Similarity:** Chuyển đổi title và description thành ma trận vector số học sử dụng kỹ thuật TF-IDF.

In [503]:
tfidf_vectorizer = TfidfVectorizer(max_features=5000)
corpus = (df_reset['title'] + ' ' + df_reset['description']).tolist()
tfidf_matrix = tfidf_vectorizer.fit_transform(corpus)

# hyperparameter để lấy top k tags
top_k = 5

tfidf_missing = tfidf_matrix[missing_idx]
tfidf_available = tfidf_matrix[available_idx]

# tính cosine similarity một lần choa tất cạ
sim_matrix = cosine_similarity(tfidf_missing, tfidf_available)

**Bước 4: Gán tags:** Gán tags cho video bị thiếu bằng cách tổng hợp tags từ 5 video giống nhất (Top-k Neighbors), ưu tiên lọc lấy các từ khóa có xuất hiện trong tiêu đề/mô tả của chính video đó.

In [504]:
for i, idx in enumerate(missing_idx):
    sim = sim_matrix[i]
    topk_idx = available_idx[np.argsort(sim)[::-1][:top_k]]
    
    # gộp tags từ neighbors
    tags_suggested = set()
    for nidx in topk_idx:
        tags_suggested.update(df_reset.at[nidx, 'tags'])
    
    # keywords từ title + description
    words_title_desc = set(df_reset.at[idx, 'title'].split() + df_reset.at[idx, 'description'].split())
    
    # intersection giữa neighbor tags và title/description
    relevant_tags = tags_suggested & words_title_desc
    if relevant_tags:
        df_reset.at[idx, 'tags'] = list(relevant_tags)
    else:
        df_reset.at[idx, 'tags'] = list(tags_suggested)

**Bước 5: Kiểm tra kết quả:** Kiểm tra kết quả trước khi gán vào df gốc

In [505]:
pd.set_option('display.max_colwidth', None)
df_reset[['title', 'tags']].head(10)

,title,tags
0,ếch ngoài đáy giếng phương mỹ chi hoá thân thành em ếch xinh em xinh say hi performance,"[htv2, vie channel, music vie channel, em xinh say hi, Em Xinh Say Hi, vie channel em xinh say hi, Em Xinh ""Say Hi"" 2025, em xinh say hi 2025 mới nhất, em xinh say hi tập 12, tập 12 em xinh, tập 12 em xinh say hi, em xinh tập 12, livestage 5 em xinh say hi, solo, solo em xinh say hi, ếch ngoài đáy giếng, ếch ngoài đáy giếng phương mỹ chi, phương mỹ chi, phương mỹ chi ếch ngoài đáy giếng, ếch ngoài đáy giếng live performance, ếch ngoài đáy giếng solo performance, phuong my chi]"
1,bài hát đang hóttrở về_như hoabài hát lấy đi nước mắt hàng triệu người con xa quêvu lan báo hiếu,"[Bài Hát hay nhất về Mẹ, Bài hát ý nghĩa nhất về Mẹ, Bài hát cảm động nhất về mẹ, Vu Lan Báo Hiếu, Như Hoa, Trở Về Như Hoa, Trở Về _ Như Hoa ( Thần Tượng Bolero 2019), Bài hát đang hót, Trở Về, Như Hoa Bolero, Như Hoa Oficial, Bài hát lấy đi nước mắt hàng triệu người con xa quê, Như Hoa Thần Tượng Bolero 2019, Bài hát đang hót nhất 2022, Mẹ Và Quê Hương]"
2,bài hát abc học bảng chữ cái abc nhạc thiếu nhi remix sôi động,"[nhạc thiếu nhi, nhạc thiếi, u nhi vui nhộn, nhạc thiếu nhi remix, bai hat thieu nhca nhac thieu nhi, ca nhạc thiếu nhi vui nhộn, nhạc thiếu nhi sôi động, nhạc thiếu nhi hay, nhạc thiếu nhi vui nhộn cho be, nhạc remix thiếu nhi, nhac tre em, nhac thieu nhi soi dong remix, ABC Song, học bảng chữ cái abc qua bài hát, bài hát học chữ cái cho bé, bài hát abc, bài hát abc tiếng việt, bé học bảng chữ cái, abcde, hoạt hình học con vật, hoạt hình học trái cây, A con cá, B con bò]"
3,nhạc cuốn là bé nhảy ngay nhảy trend tiktok thaytaichannel shorts dance,"[thay tai, thầy tài channel, thầy tài, thaytaichannel]"
4,karaoke bài ca đất phương nam tone nam nhạc sống gia huy karaoke,"[karaoke tone nữ Bài Ca Đất Phương Nam, Bài Ca Đất Phương Nam karaoke tone nam, Bài Ca Đất Phương Nam karaoke tone nữ, karaoke tone nu bai ca dat phuong nam, gia huy karaoke, karaoke nhạc trữ tình, karaoke nhạc đỏ, karaoke bolero, karaoke quan họ, karaoke nhạc sống, nhac song karaoke mien tay, nhạc sống karaoke, karaoke nhạc song, karaoke gia huy, karaokebaicadatphuongnam, bai ca dat phuong nam karaoke tone nam, gia huy, karaoke, karaoke tone nam]"
5,bài hát này xứng đáng đạt 1 tỷ view ca nhạc bolero hay nhức nách lk hỏi anh hỏi em,"[Bài hát này xứng đáng đạt 1 tỷ view, Ca nhạc bolero hay nhức nách, bolero hay nhức nách, ca nhạc bolero, hỏi anh hỏi em, lk hoi anh hoi em, thiên quang quỳnh trang, thien quang quynh trang, thiên quang, quỳnh trang, thien quang, quynh trang, bài hát tỷ view, bolero, nhạc trữ tình, nhạc vàng, bolero tru tinh, nhac vang tru tinh, thiên quang official]"
6,kèn phím melodica bài thực hành số 2 kiểm tra giũa kì hki âm nhạc 7,"[kèn phím melodica, nhạc cụ, hoa cỏ lau, phong max, Hoa Co Lau Kèn Phím Melodica, kèn phím melodica Hoa cỏ lau, HOA CỎ LAU - PHONG MAX]"
7,quả gì bài hát về động vật hay nhất liên khúc ca nhạc thiếu nhi remix vui nhộn,"[nhạc thiếu nhi, nhạc thiếu nhi vui nhộn, nhạc thiếu nhi remix, bai hat thieu nhi, ca nhac thieu nhi, ca nhạc thiếu nhi vui nhộn, nhạc thiếu nhi sôi động, nhạc thiếu nhi hay, nhạc thiếu nhi vui nhộn cho be, nhạc remix thiếu nhi, nhac tre em, nhac thieu nhi soi dong remix, nhac thieu nhi, liên khúc nhạc thiếu nhi, lien khuc thieu nhi, bài hát động vật, bài hát âm thanh động vật, động vật, bài hát những con vật đáng yêu, động vật cho bé, bài hát con vật, quả gì]"
8,nhạc thiếu nhi tập thể dục buổi sáng cá vàng bơi con heo đất bài hát sôi động cho bé ăn ngon,"[nhạc thiếu nhi tập thể dục buổi sáng, nhac thieu nhi, nhạc thiếu nhi, nhạc thiếu nhi sôi động, nhạc thiếu nhi con heo đất, nhac thieu nhi ca vang boi, tập thể dục buổi sáng, con heo đất, cá vàng bơi, bé mầm non tập thể dục, bài hát tập thể dục cho bé, tap the duc thiếu nhi, nhac thieu nhi soi dong, nhac thieu nhi hay nhat, bài hát thiếu nhi sôi động, bài hát thiếu nhi, nhạc cho bé, kenh thieu nhi bhmedia, bhmedia]"
9,bài ca sinh nhật tô gà hiha minecraft animat

**Gán `tags` đã xử lý thành công vào df gốc.**

In [506]:
df["tags"] = df_reset["tags"]

**Xử lý `search_query`**

In [507]:
#kiểm tra unique search_query
print(df['search_query'].unique())

['bài' 'mv chính thức' 'nhạc' 'nhạc không lời' 'nhạc acoustic mv'
 'nhạc bolero mv' 'nhạc buồn mv' 'nhạc chill mv' 'nhạc chế'
 'nhạc cải lương mv' 'nhạc cụ mv' 'nhạc dân ca mv' 'nhạc indie mv'
 'nhạc live mv' 'nhạc lofi mv' 'nhạc quê hương mv' 'nhạc rap mv'
 'nhạc sàn mv' 'nhạc thiếu nhi mv' 'nhạc trẻ music video'
 'nhạc trữ tình music video' 'nhạc tết mv' 'nhạc vui tươi mv'
 'nhạc vàng mv' 'nhạc đám cưới mv' 'nhạc đỏ mv' 'official'
 'OFFICIAL MUSIC VIDEO' 'remix'
 'official | MV | audio | lyric | karaoke | live | cover | remix | bài hát | nhạc']


Đây là những queries được dùng trong việc thu thập dữ liệu, sẽ không dùng cho phân tích. Tuy nhiên, để phục vụ cho mục đích bảo trì, phát triển bộ dữ liệu, ta sẽ giữ lại

### **Xử lý Binary cols (True/False)**

In [508]:
print("Giá trị duy nhất trong cột caption:")
print(df['caption'].unique())

print("Giá trị duy nhất trong cột licensed_content:")
print(df['licensed_content'].unique())

print("Giá trị duy nhất trong cột made_for_kids:")
print(df['made_for_kids'].unique())

Giá trị duy nhất trong cột caption:
[False  True]
Giá trị duy nhất trong cột licensed_content:
[ True False]
Giá trị duy nhất trong cột made_for_kids:
[False  True]


Áp dụng strip khoảng thừa cho các giá trị này:

In [509]:
# strip cách khoảng thừa của liscensed_content
df['caption'] = df['caption'].astype(str).str.strip()
df['licensed_content'] = df['licensed_content'].astype(str).str.strip()
df['made_for_kids'] = df['made_for_kids'].astype(str).str.strip()

print("Giá trị duy nhất trong cột caption sau khi strip:")
print(df['caption'].unique())

print("Giá trị duy nhất trong cột licensed_content sau khi strip:")
print(df['licensed_content'].unique())

print("Giá trị duy nhất trong cột made_for_kids sau khi strip:")
print(df['made_for_kids'].unique())

Giá trị duy nhất trong cột caption sau khi strip:
['False' 'True']
Giá trị duy nhất trong cột licensed_content sau khi strip:
['True' 'False']
Giá trị duy nhất trong cột made_for_kids sau khi strip:
['False' 'True']


### **Xử lý Datetime cols (Cột thời gian)**

**Xử lý `duration`**

Chuẩn hóa duration từ chuẩn iso về giây bình thường

In [510]:
# kiểm tra cột thời gian duration trước khi xử lý
print(df['duration'].head(10))

0       PT6M12S
1       PT6M36S
2       PT7M56S
3         PT14S
4       PT6M58S
5     PT1H45M1S
6         PT11S
7    PT1H56M15S
8      PT17M28S
9       PT3M13S
Name: duration, dtype: object


In [511]:
#duration to seconds
def iso_to_seconds(iso_str):
    pattern = re.compile(r'PT(?:(\d+)H)?(?:(\d+)M)?(?:(\d+)S)?')
    match = pattern.match(iso_str)
    if not match:
        return 0
    hours = int(match.group(1)) if match.group(1) else 0
    minutes = int(match.group(2)) if match.group(2) else 0
    seconds = int(match.group(3)) if match.group(3) else 0
    total_seconds = hours * 3600 + minutes * 60 + seconds
    return total_seconds

df['duration'] = df['duration'].apply(iso_to_seconds)

print("Giá trị trong cột duration sau khi làm sạch:")
print(df['duration'].head(10))

Giá trị trong cột duration sau khi làm sạch:
0     372
1     396
2     476
3      14
4     418
5    6301
6      11
7    6975
8    1048
9     193
Name: duration, dtype: int64


**Xử lý `published_at`**

In [512]:
# kiểm tra cột thời gian publish_at trước khi xử lý
print(df['published_at'].head(10))

0    2025-08-17T05:00:48Z
1    2022-08-24T12:00:12Z
2    2023-11-18T09:30:05Z
3    2023-07-14T03:59:08Z
4    2025-08-12T07:02:59Z
5    2025-10-04T10:15:06Z
6    2023-11-04T00:23:37Z
7    2024-10-12T11:24:17Z
8    2023-07-06T00:36:46Z
9    2021-10-27T11:30:12Z
Name: published_at, dtype: object


In [513]:
df['published_at'] = pd.to_datetime(df['published_at'], format='%Y-%m-%dT%H:%M:%SZ', errors='coerce')

print("Giá trị trong cột published_at sau khi làm sạch:")
print(df['published_at'].head(10))

Giá trị trong cột published_at sau khi làm sạch:
0   2025-08-17 05:00:48
1   2022-08-24 12:00:12
2   2023-11-18 09:30:05
3   2023-07-14 03:59:08
4   2025-08-12 07:02:59
5   2025-10-04 10:15:06
6   2023-11-04 00:23:37
7   2024-10-12 11:24:17
8   2023-07-06 00:36:46
9   2021-10-27 11:30:12
Name: published_at, dtype: datetime64[ns]


## Bước 4: Đồng bộ hóa tên cột

In [514]:
# In ra tên các cột trong DataFrame trước khi đồng bộ hóa tên cột
print("Tên các cột trước khi đồng bộ hóa:")
print(df.columns)

Tên các cột trước khi đồng bộ hóa:
Index(['video_id', 'published_at', 'channel_id', 'channel_title', 'title',
       'description', 'tags', 'category_id', 'view_count', 'like_count',
       'comment_count', 'duration', 'definition', 'caption',
       'licensed_content', 'made_for_kids', 'tags_count', 'topic_categories',
       'search_query', 'channel_subscriber_count', 'channel_view_count',
       'channel_video_count'],
      dtype='object')


In [515]:
# đổi tên title -> video_title
df.rename(columns={'title': 'video_title'}, inplace=True)
# đổi tên description -> video_description
df.rename(columns={'description': 'video_description'}, inplace=True)
# đổi tên tags -> video_tags
df.rename(columns={'tags': 'video_tags'}, inplace=True)
# đổi tên topic_categories -> video_topics
df.rename(columns={'topic_categories': 'video_topics'}, inplace=True)
# đổi tên definition -> video_definition
df.rename(columns={'definition': 'video_definition'}, inplace=True)
# đổi tên privacy_status -> video_privacy_status
df.rename(columns={'privacy_status': 'video_privacy_status'}, inplace=True)



# đổi tên category_id -> video_category_id
df.rename(columns={'category_id': 'video_category_id'}, inplace=True)
# đổi tên view_count -> video_view_count
df.rename(columns={'view_count': 'video_view_count'}, inplace=True)
# đổi tên like_count -> video_like_count
df.rename(columns={'like_count': 'video_like_count'}, inplace=True)
# đổi tên comment_count -> video_comment_count
df.rename(columns={'comment_count': 'video_comment_count'}, inplace=True)
# đổi tên tags_count -> video_tags_count
df.rename(columns={'tags_count': 'video_tags_count'}, inplace=True)


# đổi tên caption -> video_caption_status
df.rename(columns={'caption': 'video_caption_status'}, inplace=True)
# đổi tên licensed_content -> video_licensed_content
df.rename(columns={'licensed_content': 'video_licensed_content'}, inplace=True)
# đổi tên made_for_kids -> video_made_for_kids
df.rename(columns={'made_for_kids': 'video_made_for_kids'}, inplace=True)

# đổi tên published_at -> video_publish_date
df.rename(columns={'published_at': 'video_publish_date'}, inplace=True)
# đổi tên duration -> video_duration
df.rename(columns={'duration': 'video_duration'}, inplace=True)

In [516]:
# Tên các cột sau khi đồng bộ hóa
print("Tên các cột sau khi đồng bộ hóa:")
print(df.columns)

Tên các cột sau khi đồng bộ hóa:
Index(['video_id', 'video_publish_date', 'channel_id', 'channel_title',
       'video_title', 'video_description', 'video_tags', 'video_category_id',
       'video_view_count', 'video_like_count', 'video_comment_count',
       'video_duration', 'video_definition', 'video_caption_status',
       'video_licensed_content', 'video_made_for_kids', 'video_tags_count',
       'video_topics', 'search_query', 'channel_subscriber_count',
       'channel_view_count', 'channel_video_count'],
      dtype='object')


## Kiểm tra lại trước khi bước vào EDA

In [517]:
df.head(5)

,video_id,video_publish_date,channel_id,channel_title,video_title,video_description,video_tags,video_category_id,video_view_count,video_like_count,...,video_definition,video_caption_status,video_licensed_content,video_made_for_kids,video_tags_count,video_topics,search_query,channel_subscriber_count,channel_view_count,channel_video_count
0,sZrIbpwjTwk,2025-08-17 05:00:48,UC2fu6CiFfNYz5UFORvFyc0w,Vie Channel - MUSIC,ếch ngoài đáy giếng phương mỹ chi hoá thân thành em ếch xinh em xinh say hi performance,ếch ngoài đáy giếng phương mỹ chi hoá thân thành em ếch xinh em xinh say hi performance emxinhsayhi emxinh echngoaidaygieng phuongmychi emxinhsayhi dautruonggiatoc viechannelmusic viechannelmusic viechannel vieon datvietvac 30namcomotdatviet xem ngay tập full httpswwwyoutubecomwatchvo_lqxbes_s xem sớm nhất trên ứng dụng vieon httpsvieonvnphimhaymtvshowgamethucte xem thêm tại đây httpsexshbfanlinkexshep12 đăng ký hội viên ngay nào httpswwwyoutubecomchanneluckna2ocun1e6u5i8gvtdkowjoin mọi nhu cầu quảng cáo và tài trợ xin liên hệ công ty cổ phần vie channel hotline 0907950371 mr đức khoa email khoanguyenviechannelvn hoặc salesviechannelvn đừng quên subscribe để thưởng thức thật nhiều bài hát hay nhất nhé subscribe ngay httpswwwyoutubecomchanneluc2fu6ciffnyz5uforvfyc0wsub_confirmation1 tải app vieon tại httpsclickvieonvnwsbgytapp xem các chương trình hấp dẫn khác của vie channel em xinh say hi full httpsbitlyemxinhsayhi2025 em xinh say hi performance httpsbitlyemxinhsayhiperformance anh trai say hi full httpsbitlyanhtraisayhi anh trai say hi performance httpsbitlyanhtraisayhiperformance rap việt full httpsbitlyrapvietss3 rap việt performance httpsbitlyrapviet2024performance hoặc xem thêm trên hệ thống vie channel website httpviechannelvn fanpage httpswwwfacebookcomviechannelhtv2 instagram httpswwwinstagramcomviechannelhtv2,"[htv2, vie channel, music vie channel, em xinh say hi, Em Xinh Say Hi, vie channel em xinh say hi, Em Xinh ""Say Hi"" 2025, em xinh say hi 2025 mới nhất, em xinh say hi tập 12, tập 12 em xinh, tập 12 em xinh say hi, em xinh tập 12, livestage 5 em xinh say hi, solo, solo em xinh say hi, ếch ngoài đáy giếng, ếch ngoài đáy giếng phương mỹ chi, phương mỹ chi, phương mỹ chi ếch ngoài đáy giếng, ếch ngoài đáy giếng live performance, ếch ngoài đáy giếng solo performance, phuong my chi]",10,82407446,234808.0,...,hd,False,True,False,22,"[Music, Music of Asia, Pop music, Television program]",bài,1710000.0,2.303825e+09,2519.0
1,Qgzb0musH3A,2022-08-24 12:00:12,UCIKpxDD9CUS7vBCJte7l-UQ,Như Hoa Official,bài hát đang hóttrở về_như hoabài hát lấy đi nước mắt hàng triệu người con xa quêvu lan báo hiếu,bài hát đang hót trở về như hoa thần tượng bolero 2019 bài hát lấy đi nước mắt hàng triệu người con xa quê vu lan báo hiếu 2022 mv official composer tuấn phương singer như hoa thần tượng bolero 2019 record mix võ hoàng studio directors như hoa official produced by khang hy entertainment lyric trở về sáng tác nhạc sĩ tuấn phương ta trở về trong tiếng gọi mẹ ơi về với mẹ về nơi thật yên bình nước mắt ta rơi bởi một điều có thật rồi một ngày mẹ sẽ đi xa ta trở về bên cánh đồng quê xưa hạt lúa vàng hạt mồ hôi chát mặn ta lớn khôn thêm giữa nghĩa tình xóm làng chia cho nhau ngụm nước giữa trưa hè đk ta về đây khu vườn rộn tiếng ve có cánh diều mang tuổi thơ khát vọng có cánh cò bên nôi mẹ hát có hàng tre ru chiều gió thênh thang nếu cuộc sống ban cho ta điều ước ta xin được về những năm tháng tuổi thơ nghe lá hát trên từng con đường nhỏ tổ quốc mình chẳng thể nào thay vẫn còn đây tím chiều nhuộm áo em vẫn câu hò như lời ai nhắn nhủ vẫn con đò mang bao niềm nhớ vẫn mẹ ta bên thềm ngóng con xa nếu cuộc sống ban cho ta điều ước ta xin được về những năm tháng tuổi thơ nghe lá hát trên từng con đường nhỏ tổ quốc mình chẳng thể nào thay __________________________ đây là kênh you tobe chính thức của ca sĩ như hoa official thần tượng bolero 2019 cảm ơn các bạn đã quan tâm theo dõi kênh ca nhạc của như hoa chúc quý vị có những giây phút nghe nhạc t

In [518]:
# Lưu kêt quả đã làm sạch
df.to_csv("../data/youtube_vn_music_cleaned.csv", index=False)